File: npm/packages/rvforge/src/manifest.ts
/**
 * Canonical build manifest generation (ADR-283 §7, RVM integration contract).
 *
 * The manifest is the thing a build is traceable to, so it is deterministic:
 * sorted keys, sorted target and capability lists, and no timestamps, paths,
 * or hostnames. The same logical build description always serialises to the
 * same bytes and therefore the same `manifestSha256`. Anything that varies per
 * run — the clock, the builder, the machine — belongs in the provenance
 * record instead.
 *
 * The manifest carries signing *references* only. Key material never enters
 * it, is never logged, and is never transmitted.
 */

import { ForgeError, ForgeErrorCode } from './errors';
import { canonicalJson, canonicalJsonFile, sha256Canonical } from './hash';
import {
  CAPABILITY_CLASSES,
  PACKAGING_MODES,
  RUNTIME_PROFILES,
  resolveTargets,
  type AppMetadata,
  type BuildManifest,
  type CapabilityClass,
  type CapabilityPolicy,
  type PackagingMode,
  type RuntimeProfile,
  type RvfIdentityRecord,
  type SigningRefs,
} from './types';
import { FORGE_PACKAGE_NAME, forgeVersion } from './version';

/**
 * Runtime profiles present in the published RVM compatibility matrix.
 *
 * The MVP ships WASM execution only (ADR-283 §8); native acceleration, Linux
 * microVM, and bare-metal RVM are deferred. Forge rejects a combination that
 * is absent from the matrix rather than emitting a package whose runtime
 * contract nothing can honour (RVM §9).
 */
export const RVM_COMPATIBILITY_MATRIX: Readonly<Record<RuntimeProfile, boolean>> = Object.freeze({
  wasm: true,
  native: false,
  microvm: false,
  rvm: false,
});

const SEMVER = /^\d+\.\d+\.\d+(?:[-+][0-9A-Za-z.-]+)*$/;
const IDENTIFIER = /^[A-Za-z0-9]([A-Za-z0-9-]*[A-Za-z0-9])?(\.[A-Za-z0-9]([A-Za-z0-9-]*[A-Za-z0-9])?)+$/;

export interface BuildManifestInput {
  app: AppMetadata;
  identity: RvfIdentityRecord;
  targets: readonly string[];
  packaging: { mode: PackagingMode; updateChannel?: string; distributionUrl?: string };
  runtime: { profile: RuntimeProfile; rvmVersion: string; rvmCommit: string };
  capabilityPolicy?: Partial<CapabilityPolicy>;
  signing?: SigningRefs;
  /** Overrides the detected generator version; used by tests for determinism. */
  generatorVersion?: string;
}

/** Default-deny: every capability class present, every allowlist empty. */
export function defaultCapabilityPolicy(): CapabilityPolicy {
  const allow: Partial<Record<CapabilityClass, string[]>> = {};
  for (const cls of CAPABILITY_CLASSES) allow[cls] = [];
  return { defaultDeny: true, allow };
}

/**
 * Normalise a policy into canonical form: `defaultDeny` is forced true, every
 * class is present, and each allowlist is de-duplicated and sorted.
 *
 * Forcing `defaultDeny` is deliberate. A policy that opts out of default-deny
 * is not a policy forge knows how to express, and silently accepting one would
 * put an undeclared-capability grant into a signed manifest.
 */
export function normalizeCapabilityPolicy(policy?: Partial<CapabilityPolicy>): CapabilityPolicy {
  const normalized = defaultCapabilityPolicy();
  if (!policy?.allow) return normalized;

  for (const [rawClass, rawGrants] of Object.entries(policy.allow)) {
    if (!(CAPABILITY_CLASSES as readonly string[]).includes(rawClass)) {
      throw new ForgeError(
        ForgeErrorCode.MANIFEST,
        `Unknown capability class "${rawClass}".`,
        { capabilityClass: rawClass, known: [...CAPABILITY_CLASSES] },
      );
    }
    if (!Array.isArray(rawGrants)) {
      throw new ForgeError(
        ForgeErrorCode.MANIFEST,
        `Capability class "${rawClass}" must map to an array of grants.`,
        { capabilityClass: rawClass },
      );
    }
    for (const grant of rawGrants) {
      if (typeof grant !== 'string' || grant.trim() === '') {
        throw new ForgeError(
          ForgeErrorCode.MANIFEST,
          `Capability class "${rawClass}" contains a non-string or empty grant.`,
          { capabilityClass: rawClass },
        );
      }
    }
    normalized.allow[rawClass as CapabilityClass] = [...new Set(rawGrants.map((g) => g.trim()))].sort();
  }
  return normalized;
}

/** SHA256 over the canonical JSON of a normalised capability policy. */
export const capabilityPolicyHash = (policy: CapabilityPolicy): string => sha256Canonical(policy);

/**
 * Build the canonical manifest.
 *
 * @throws {ForgeError} `FORGE_E_MANIFEST` for invalid metadata, an unsupported
 * runtime combination, or a malformed capability policy;
 * `FORGE_E_UNSUPPORTED_TARGET` for an unknown target.
 */
export function buildManifest(input: BuildManifestInput): BuildManifest {
  const app = validateApp(input.app);
  const targets = resolveTargets(input.targets);
  if (targets.length === 0) {
    throw new ForgeError(ForgeErrorCode.MANIFEST, 'At least one build target is required.');
  }

  if (!PACKAGING_MODES.includes(input.packaging.mode)) {
    throw new ForgeError(
      ForgeErrorCode.MANIFEST,
      `Unknown packaging mode "${input.packaging.mode}".`,
      { mode: input.packaging.mode, known: [...PACKAGING_MODES] },
    );
  }
  if (input.packaging.mode === 'thin' && !input.packaging.distributionUrl) {
    throw new ForgeError(
      ForgeErrorCode.MANIFEST,
      'Thin packaging requires packaging.distributionUrl — the reader has to know where to fetch the signed RVF.',
    );
  }

  validateRuntime(input.runtime);
  validateIdentity(input.identity);

  const capabilityPolicy = normalizeCapabilityPolicy(input.capabilityPolicy);

  return {
    schemaVersion: 1,
    generator: { name: FORGE_PACKAGE_NAME, version: input.generatorVersion ?? forgeVersion() },
    app,
    rvf: {
      sha256: input.identity.sha256,
      size: input.identity.size,
      fileId: input.identity.fileId,
      formatVersion: input.identity.formatVersion,
      signatureAlgo: input.identity.signatureAlgo,
      signaturePresent: input.identity.signaturePresent,
    },
    targets,
    packaging: {
      mode: input.packaging.mode,
      ...(input.packaging.updateChannel ? { updateChannel: input.packaging.updateChannel } : {}),
      ...(input.packaging.distributionUrl ? { distributionUrl: input.packaging.distributionUrl } : {}),
    },
    runtime: {
      rvmVersion: input.runtime.rvmVersion,
      rvmCommit: input.runtime.rvmCommit,
      runtimeProfile: input.runtime.profile,
      stateSchemaVersion: 1,
      witnessSchemaVersion: 1,
    },
    capabilityPolicy,
    capabilityPolicyHash: capabilityPolicyHash(capabilityPolicy),
    signing: normalizeSigning(input.signing),
  };
}

function validateApp(app: AppMetadata): AppMetadata {
  requireText(app.name, 'app.name');
  requireText(app.publisher, 'app.publisher');
  if (!SEMVER.test(app.version)) {
    throw new ForgeError(
      ForgeErrorCode.MANIFEST,
      `app.version "${app.version}" is not a semantic version.`,
      { version: app.version },
    );
  }
  if (!IDENTIFIER.test(app.identifier)) {
    throw new ForgeError(
      ForgeErrorCode.MANIFEST,
      `app.identifier "${app.identifier}" must be reverse-DNS, e.g. "io.ruv.myagent".`,
      { identifier: app.identifier },
    );
  }
  return {
    name: app.name.trim(),
    version: app.version.trim(),
    publisher: app.publisher.trim(),
    identifier: app.identifier.trim(),
    ...(app.description ? { description: app.description.trim() } : {}),
    ...(app.icon ? { icon: app.icon.trim() } : {}),
  };
}

function validateRuntime(runtime: BuildManifestInput['runtime']): void {
  if (!RUNTIME_PROFILES.includes(runtime.profile)) {
    throw new ForgeError(
      ForgeErrorCode.MANIFEST,
      `Unknown runtime profile "${runtime.profile}".`,
      { profile: runtime.profile, known: [...RUNTIME_PROFILES] },
    );
  }
  if (!RVM_COMPATIBILITY_MATRIX[runtime.profile]) {
    throw new ForgeError(
      ForgeErrorCode.MANIFEST,
      `Runtime profile "${runtime.profile}" is absent from the published RVM compatibility matrix; this release packages "wasm" only.`,
      { profile: runtime.profile, supported: ['wasm'] },
    );
  }
  if (!SEMVER.test(runtime.rvmVersion)) {
    throw new ForgeError(
      ForgeErrorCode.MANIFEST,
      `runtime.rvmVersion "${runtime.rvmVersion}" is not a semantic version.`,
      { rvmVersion: runtime.rvmVersion },
    );
  }
  requireText(runtime.rvmCommit, 'runtime.rvmCommit');
}

function validateIdentity(identity: RvfIdentityRecord): void {
  if (!/^[0-9a-f]{64}$/.test(identity.sha256)) {
    throw new ForgeError(
      ForgeErrorCode.MANIFEST,
      'rvf.sha256 must be a 64-character lowercase hex digest. Run validation with --deep to compute it.',
    );
  }
  if (!Number.isInteger(identity.size) || identity.size <= 0) {
    throw new ForgeError(ForgeErrorCode.MANIFEST, 'rvf.size must be a positive integer.');
  }
  if (!/^[0-9a-f]{32}$/.test(identity.fileId)) {
    throw new ForgeError(ForgeErrorCode.MANIFEST, 'rvf.fileId must be a 32-character lowercase hex value.');
  }
}

function normalizeSigning(signing?: SigningRefs): SigningRefs {
  if (!signing) return {};
  const out: SigningRefs = {};
  for (const platform of ['linux', 'macos', 'windows'] as const) {
    const ref = signing[platform];
    if (!ref) continue;
    requireText(ref.provider, `signing.${platform}.provider`);
    requireText(ref.keyRef, `signing.${platform}.keyRef`);
    out[platform] = {
      provider: ref.provider.trim(),
      keyRef: ref.keyRef.trim(),
      ...(ref.identity ? { identity: ref.identity.trim() } : {}),
    };
  }
  return out;
}

function requireText(value: unknown, field: string): void {
  if (typeof value !== 'string' || value.trim() === '') {
    throw new ForgeError(ForgeErrorCode.MANIFEST, `${field} is required and must be a non-empty string.`, {
      field,
    });
  }
}

/** Canonical JSON for a manifest (no trailing newline). */
export const serializeManifest = (manifest: BuildManifest): string => canonicalJson(manifest);

/** Exact bytes forge writes to `build-manifest.json`. */
export const manifestFileContents = (manifest: BuildManifest): string => canonicalJsonFile(manifest);

/** SHA256 over the canonical manifest encoding — the build's manifest identity. */
export const manifestHash = (manifest: BuildManifest): string => sha256Canonical(manifest);

File: crates/rvf-forge-core/src/manifest.rs
//! The canonical build manifest and the runtime contract it carries.
//!
//! ADR-291 §1 fixes seven fields that every generated package must embed. The
//! build manifest is the superset: it adds the application metadata, targets,
//! and packaging mode that describe *how* the package is produced, and it is
//! what the CLI signs and submits.
//!
//! Two properties make it useful rather than decorative:
//!
//! - **Determinism.** [`CanonicalBuildManifest::to_canonical_json`] is the
//!   hashed form, and the same [`BuildSpec`] always produces the same bytes.
//!   Without that, `buildManifestHash` in a provenance record identifies
//!   nothing.
//! - **Gating before compute.** [`build_manifest`] rejects targets absent from
//!   the compatibility matrix (ADR-291 §2) at manifest-generation time, before
//!   upload and before a worker is allocated.

use crate::canonical::{canonical_sha256_hex, to_canonical_json};
use crate::capability::CapabilityPolicy;
use crate::error::{ForgeError, Result};
use crate::target::Target;
use serde::{Deserialize, Serialize};

/// Schema identifier carried by every [`CanonicalBuildManifest`].
pub const BUILD_MANIFEST_SCHEMA: &str = "rvforge.build-manifest/1";

/// Schema identifier carried by every [`RuntimeContract`].
pub const RUNTIME_CONTRACT_SCHEMA: &str = "rvforge.runtime-contract/1";

/// The state delta and checkpoint schema this crate emits (ADR-291 §1).
pub const STATE_SCHEMA_VERSION: u32 = 1;

/// The witness record schema this crate emits (ADR-291 §1).
pub const WITNESS_SCHEMA_VERSION: u32 = 1;

/// How the RVF reaches the installed machine (ADR-283 §3).
#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum PackagingMode {
    /// The complete RVF plus the reader ship in every installer. Runs with no
    /// network access; duplicates large models across outputs.
    Embedded,
    /// The reader plus a signed RVF locator; the reader downloads and verifies
    /// before execution.
    Thin,
    /// The reader is installed once and `.rvf` is registered as a file type.
    SharedReader,
}

/// The runtime family a package was built and validated for (ADR-291 §1).
///
/// The reader may select a runtime at or below the profile a package was built
/// for and never above it (ADR-291 §4), which is what [`RuntimeProfile::rank`]
/// encodes.
#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum RuntimeProfile {
    /// WebAssembly execution.
    Wasm,
    /// A Linux microVM.
    LinuxMicrovm,
    /// Native execution under OS isolation.
    Native,
    /// Native RVM, the strongest profile.
    Rvm,
}

impl RuntimeProfile {
    /// Position in the runtime preference order of ADR-291 §4, strongest last.
    pub const fn rank(self) -> u8 {
        match self {
            Self::Wasm => 0,
            Self::LinuxMicrovm => 1,
            Self::Native => 2,
            Self::Rvm => 3,
        }
    }

    /// Whether a package built for `self` may select `candidate` at run time.
    ///
    /// A package built for the WASM profile never promotes itself to native
    /// RVM (ADR-291 acceptance criterion 9).
    pub const fn permits(self, candidate: Self) -> bool {
        candidate.rank() <= self.rank()
    }
}

/// Publisher-facing metadata that brands the produced packages (FR007).
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppMetadata {
    /// Application name shown to the installing user.
    pub name: String,
    /// Application version.
    pub version: String,
    /// Publisher identity string.
    pub publisher: String,
    /// SPDX license identifier or a license name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub license: Option<String>,
    /// Short description.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

/// What the caller asks Forge to build.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct BuildSpec {
    /// SHA-256 of the canonical RVF, lowercase hex — normally
    /// [`crate::ForgeInspection::rvf_identity`].
    pub rvf_identity: String,
    /// Application branding.
    pub app: AppMetadata,
    /// Targets to build. Order does not affect the manifest; duplicates are
    /// collapsed.
    pub targets: Vec<Target>,
    /// Packaging mode.
    pub packaging_mode: PackagingMode,
    /// Runtime family this build is validated for.
    pub runtime_profile: RuntimeProfile,
    /// The default-deny capability policy this build is made against.
    pub capability_policy: CapabilityPolicy,
    /// Semantic version of the embedded RVM.
    pub rvm_version: String,
    /// Source revision of the embedded RVM. Required: a semantic version alone
    /// does not identify a build (ADR-291, Alternatives Considered).
    pub rvm_commit: String,
    /// Revision of the compatibility matrix consulted, recorded so a past
    /// admission decision can be reconstructed (ADR-291 §2).
    pub compatibility_matrix_revision: Option<String>,
}

/// The seven-field contract embedded in every generated package (ADR-291 §1).
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RuntimeContract {
    /// Schema identifier; always [`RUNTIME_CONTRACT_SCHEMA`].
    pub schema: String,
    /// SHA-256 of the canonical RVF. Identical across every package from one
    /// build, and unchanged from the submitted artifact.
    pub rvf_identity: String,
    /// Semantic version of the embedded runtime.
    pub rvm_version: String,
    /// Source revision of the embedded runtime.
    pub rvm_commit: String,
    /// Runtime family this package was built for.
    pub runtime_profile: RuntimeProfile,
    /// SHA-256 of the capability policy this package was built against.
    pub capability_policy_hash: String,
    /// State delta and checkpoint schema version.
    pub state_schema_version: u32,
    /// Witness record schema version.
    pub witness_schema_version: u32,
}

/// The full build manifest.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CanonicalBuildManifest {
    /// Schema identifier; always [`BUILD_MANIFEST_SCHEMA`].
    pub schema: String,
    /// SHA-256 of the canonical RVF, lowercase hex.
    pub rvf_identity: String,
    /// Application branding.
    pub app: AppMetadata,
    /// Targets to build, deduplicated and sorted so that two specs requesting
    /// the same set produce the same manifest.
    pub targets: Vec<Target>,
    /// Packaging mode.
    pub packaging_mode: PackagingMode,
    /// Runtime family this build is validated for.
    pub runtime_profile: RuntimeProfile,
    /// SHA-256 of the capability policy.
    pub capability_policy_hash: String,
    /// Semantic version of the embedded runtime.
    pub rvm_version: String,
    /// Source revision of the embedded runtime.
    pub rvm_commit: String,
    /// State delta and checkpoint schema version; always
    /// [`STATE_SCHEMA_VERSION`].
    pub state_schema_version: u32,
    /// Witness record schema version; always [`WITNESS_SCHEMA_VERSION`].
    pub witness_schema_version: u32,
    /// Revision of the compatibility matrix that admitted this build.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub compatibility_matrix_revision: Option<String>,
}

impl CanonicalBuildManifest {
    /// The canonical JSON encoding: sorted keys, no floats, no whitespace.
    ///
    /// # Errors
    ///
    /// [`ForgeError::Manifest`] if the manifest cannot be canonicalized.
    pub fn to_canonical_json(&self) -> Result<String> {
        to_canonical_json(self)
    }

    /// SHA-256 of the canonical JSON, lowercase hex. This is
    /// `buildManifestHash` in a provenance record.
    ///
    /// # Errors
    ///
    /// As [`CanonicalBuildManifest::to_canonical_json`].
    pub fn manifest_hash(&self) -> Result<String> {
        canonical_sha256_hex(self)
    }

    /// The seven-field runtime contract this manifest implies.
    pub fn runtime_contract(&self) -> RuntimeContract {
        RuntimeContract {
            schema: RUNTIME_CONTRACT_SCHEMA.to_string(),
            rvf_identity: self.rvf_identity.clone(),
            rvm_version: self.rvm_version.clone(),
            rvm_commit: self.rvm_commit.clone(),
            runtime_profile: self.runtime_profile,
            capability_policy_hash: self.capability_policy_hash.clone(),
            state_schema_version: self.state_schema_version,
            witness_schema_version: self.witness_schema_version,
        }
    }
}

/// Build a canonical manifest from `spec`.
///
/// Validation happens here, before upload and before a worker is allocated
/// (ADR-291 §2): the RVF identity must be a 64-character hex SHA-256, at least
/// one target must be requested, every target must be in the compatibility
/// matrix, and `rvmVersion` and `rvmCommit` must both be present.
///
/// # Errors
///
/// [`ForgeError::UnsupportedTarget`] for a target absent from the matrix;
/// [`ForgeError::Manifest`] for every other validation failure.
pub fn build_manifest(spec: &BuildSpec) -> Result<CanonicalBuildManifest> {
    validate_identity(&spec.rvf_identity)?;
    require_non_empty("app.name", &spec.app.name)?;
    require_non_empty("app.version", &spec.app.version)?;
    require_non_empty("app.publisher", &spec.app.publisher)?;
    require_non_empty("rvmVersion", &spec.rvm_version)?;
    require_non_empty("rvmCommit", &spec.rvm_commit)?;

    if spec.targets.is_empty() {
        return Err(ForgeError::Manifest {
            detail: "no targets requested; a build must name at least one target".to_string(),
        });
    }
    for target in &spec.targets {
        target.ensure_supported()?;
    }

    // Deduplicate and sort so that target order at the call site cannot change
    // the manifest bytes.
    let mut targets = spec.targets.clone();
    targets.sort_unstable();
    targets.dedup();

    Ok(CanonicalBuildManifest {
        schema: BUILD_MANIFEST_SCHEMA.to_string(),
        rvf_identity: spec.rvf_identity.to_ascii_lowercase(),
        app: spec.app.clone(),
        targets,
        packaging_mode: spec.packaging_mode,
        runtime_profile: spec.runtime_profile,
        capability_policy_hash: spec.capability_policy.policy_hash()?,
        rvm_version: spec.rvm_version.clone(),
        rvm_commit: spec.rvm_commit.clone(),
        state_schema_version: STATE_SCHEMA_VERSION,
        witness_schema_version: WITNESS_SCHEMA_VERSION,
        compatibility_matrix_revision: spec.compatibility_matrix_revision.clone(),
    })
}

fn validate_identity(identity: &str) -> Result<()> {
    if identity.len() != 64 || !identity.chars().all(|c| c.is_ascii_hexdigit()) {
        return Err(ForgeError::Manifest {
            detail: format!(
                "rvfIdentity must be a 64-character hex SHA-256, got {} characters",
                identity.len()
            ),
        });
    }
    Ok(())
}

fn require_non_empty(field: &str, value: &str) -> Result<()> {
    if value.trim().is_empty() {
        return Err(ForgeError::Manifest {
            detail: format!("{field} must not be empty"),
        });
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::capability::{CapabilityClass, CapabilityGrant};

    fn spec() -> BuildSpec {
        BuildSpec {
            rvf_identity: "ab".repeat(32),
            app: AppMetadata {
                name: "Example Agent".into(),
                version: "1.2.3".into(),
                publisher: "Example Inc".into(),
                license: Some("MIT".into()),
                description: None,
            },
            targets: vec![Target::WindowsX64, Target::MacosUniversal],
            packaging_mode: PackagingMode::Embedded,
            runtime_profile: RuntimeProfile::Wasm,
            capability_policy: CapabilityPolicy::deny_all(),
            rvm_version: "0.4.0".into(),
            rvm_commit: "abc123".into(),
            compatibility_matrix_revision: Some("matrix-2026-08-01".into()),
        }
    }

    #[test]
    fn manifest_carries_the_adr_291_contract_fields() {
        let m = build_manifest(&spec()).unwrap();
        assert_eq!(m.schema, BUILD_MANIFEST_SCHEMA);
        assert_eq!(m.state_schema_version, 1);
        assert_eq!(m.witness_schema_version, 1);
        assert_eq!(m.rvm_version, "0.4.0");
        assert_eq!(m.rvm_commit, "abc123");
        assert_eq!(m.capability_policy_hash.len(), 64);

        let contract = m.runtime_contract();
        assert_eq!(contract.rvf_identity, m.rvf_identity);
        assert_eq!(contract.runtime_profile, RuntimeProfile::Wasm);
        let json = serde_json::to_string(&contract).unwrap();
        for field in [
            "rvfIdentity",
            "rvmVersion",
            "rvmCommit",
            "runtimeProfile",
            "capabilityPolicyHash",
            "stateSchemaVersion",
            "witnessSchemaVersion",
        ] {
            assert!(json.contains(field), "contract is missing {field}: {json}");
        }
    }

    #[test]
    fn the_same_spec_produces_byte_identical_json() {
        let a = build_manifest(&spec())
            .unwrap()
            .to_canonical_json()
            .unwrap();
        let b = build_manifest(&spec())
            .unwrap()
            .to_canonical_json()
            .unwrap();
        assert_eq!(a, b);
    }

    #[test]
    fn target_order_does_not_change_the_manifest() {
        let mut reversed = spec();
        reversed.targets.reverse();
        assert_eq!(
            build_manifest(&spec()).unwrap().manifest_hash().unwrap(),
            build_manifest(&reversed).unwrap().manifest_hash().unwrap()
        );
    }

    #[test]
    fn duplicate_targets_are_collapsed() {
        let mut dupes = spec();
        dupes.targets = vec![
            Target::WindowsX64,
            Target::WindowsX64,
            Target::MacosUniversal,
        ];
        assert_eq!(
            build_manifest(&dupes).unwrap().targets,
            build_manifest(&spec()).unwrap().targets
        );
    }

    #[test]
    fn a_different_capability_policy_changes_the_manifest_hash() {
        let mut opened = spec();
        opened.capability_policy = CapabilityPolicy::deny_all()
            .grant(CapabilityClass::Network, CapabilityGrant::unscoped());
        assert_ne!(
            build_manifest(&spec()).unwrap().manifest_hash().unwrap(),
            build_manifest(&opened).unwrap().manifest_hash().unwrap()
        );
    }

    #[test]
    fn canonical_json_has_sorted_keys_and_no_separator_whitespace() {
        let json = build_manifest(&spec())
            .unwrap()
            .to_canonical_json()
            .unwrap();
        // Whitespace inside string values is content; whitespace around the
        // structural characters is what canonicalization removes.
        assert!(!json.contains("\": "), "{json}");
        assert!(!json.contains(", \""), "{json}");
        assert!(!json.contains('\n'), "{json}");

        let keys = [
            "app",
            "capabilityPolicyHash",
            "rvfIdentity",
            "schema",
            "targets",
        ];
        let positions: Vec<usize> = keys
            .iter()
            .map(|k| {
                json.find(&format!("\"{k}\""))
                    .unwrap_or_else(|| panic!("missing {k}"))
            })
            .collect();
        assert!(
            positions.windows(2).all(|w| w[0] < w[1]),
            "keys are not in sorted order: {json}"
        );
    }

    #[test]
    fn identity_is_lowercased() {
        let mut upper = spec();
        upper.rvf_identity = "AB".repeat(32);
        assert_eq!(
            build_manifest(&upper).unwrap().rvf_identity,
            "ab".repeat(32)
        );
    }

    #[test]
    fn a_bad_identity_is_rejected() {
        for bad in ["", "abc", &"z".repeat(64)] {
            let mut s = spec();
            s.rvf_identity = bad.to_string();
            let err = build_manifest(&s).unwrap_err();
            assert_eq!(err.code(), crate::ErrorCode::Manifest, "{bad}");
        }
    }

    #[test]
    fn empty_required_fields_are_rejected() {
        let mut s = spec();
        s.rvm_commit = "  ".into();
        let err = build_manifest(&s).unwrap_err();
        assert_eq!(err.code(), crate::ErrorCode::Manifest);
        assert!(err.to_string().contains("rvmCommit"), "{err}");
    }

    #[test]
    fn no_targets_is_rejected() {
        let mut s = spec();
        s.targets.clear();
        assert_eq!(
            build_manifest(&s).unwrap_err().code(),
            crate::ErrorCode::Manifest
        );
    }

    #[test]
    fn runtime_profile_never_promotes_above_the_built_profile() {
        assert!(RuntimeProfile::Wasm.permits(RuntimeProfile::Wasm));
        assert!(!RuntimeProfile::Wasm.permits(RuntimeProfile::Rvm));
        assert!(!RuntimeProfile::Wasm.permits(RuntimeProfile::Native));
        assert!(RuntimeProfile::Rvm.permits(RuntimeProfile::Wasm));
        assert!(RuntimeProfile::Native.permits(RuntimeProfile::LinuxMicrovm));
    }
}
